### PR TITLE
Added checks for description and longDescription length, description being non-empty

### LINKED ISSUE #332: zimwriterfs should allow the new `--longDescription`
Look at https://wiki.openzim.org/wiki/Metadata

Would be great if `--description` and `--longDescription` max length are checked properly as well.

### PR DESCRIPTION
**Pull request by Onyx2406**
fixes #332 and #336

### CODE REVIEW

**Comment by kelson42** [comment]
@Onyx2406 In case this is not clear, you don't have implemented everything requested.

**Comment by Onyx2406** [comment]
can you review this now? @kelson42 

I made some commits

**Comment by Onyx2406** [comment]
Done @kelson42

**Review by kelson42 — APPROVED** [review]
LGTM, requesting @veloman-yunkan for a more technical review.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
Please squash all commits into one.

File: src/zimwriterfs/zimwriterfs.cpp [hunk 1]
```diff
@@ -296,6 +303,16 @@ void parse_args(int argc, char** argv)
...
+  
+  if (description.length() > 80) {
+  std::cerr << "Description length exceeds the 80 character limit." << std::endl;
+    exit(1);
+  }
+
+  if (!longDescription.empty() && longDescription.length() > 4000) {
+    std::cerr << "Long description length exceeds the 4000 character limit." << std::endl;
+    exit(1);
+  }
```
[the remark this answers is not in the document]
Comment (kelson42) [reply src/zimwriterfs/zimwriterfs.cpp:315]: Sounds a smart idea, at least as warning.

[the remark this answers is not in the document]
Comment (Onyx2406) [reply src/zimwriterfs/zimwriterfs.cpp:315]: should I implement this?

[the remark this answers is not in the document]
Comment (Onyx2406) [reply src/zimwriterfs/zimwriterfs.cpp:315]: > should I implement this?

if it is shorter, then what should I print? and what should it do?

File: src/zimwriterfs/zimwriterfs.cpp [hunk 2]
```diff
@@ -189,6 +192,7 @@ void parse_args(int argc, char** argv)
       = {{"help", no_argument, 0, 'h'},
          {"verbose", no_argument, 0, 'v'},
          {"version", no_argument, 0, 'V'},
+         {"longDescription", required_argument, 0, 'L'},
```
[the remark this answers is not in the document]
Comment (Onyx2406) [reply src/zimwriterfs/zimwriterfs.cpp:195]: can you tell how can I fix this? Do I just remove line 195 or it should be no_argument?

[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/zimwriterfs/zimwriterfs.cpp:195]: Oops, I misinterpreted the meaning of `required_argument`. Nothing to fix here.

File: src/zimwriterfs/zimwriterfs.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kelson42) [reply src/zimwriterfs/zimwriterfs.cpp:315]: Yes, take initiative

**Comment by Onyx2406** [comment]
Done @kelson42 , can you review?

**Comment by kelson42** [comment]
@Onyx2406 I have already approved, @veloman-yunkan is your reviewer now. You don't need to put a comment, just click on the review rerequesting icon.

File: src/zimwriterfs/zimwriterfs.cpp [hunk 3]
```diff
@@ -80,6 +81,11 @@ bool thereAreMissingArguments()
       || illustration.empty();
 }
 
+bool checkDescriptionLengths() {
+  if (!longDescription.empty() && longDescription.length() < description.length()) {
+    return true;
+  }
+  return false;
 }
```
Comment (veloman-yunkan) [inline src/zimwriterfs/zimwriterfs.cpp:89]: Such functions usually return `false` when the check fails.

File: src/zimwriterfs/zimwriterfs.cpp [hunk 4]
```diff
@@ -296,6 +308,21 @@ void parse_args(int argc, char** argv)
...
+  
+  if (description.length() > 80) {
+    std::cerr << "Description length exceeds the 80 character limit." << std::endl;
+    exit(1);
+  }
+
+  if (!longDescription.empty() && longDescription.length() > 4000) {
+    std::cerr << "Long description length exceeds the 4000 character limit." << std::endl;
+    exit(1);
+  }
```
Comment (veloman-yunkan) [inline src/zimwriterfs/zimwriterfs.cpp:325]: 1. This (except the call to `exit()` can be moved inside the `checkDescriptionLengths()` function
2. Please also add a check on description being non-empty (that will fix #336).

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
The commit description must be edited

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
Code is good. However the commit message body contains some text that has little to no value. Once the commit message is fixed this PR can be merged.

Note that [std::string::length()](https://en.cppreference.com/w/cpp/string/basic_string/length) returns the size of the string in bytes rather than in characters, i.e. the checks will not work correctly on non-ascii texts. But that can be fixed in s separate PR.

**Comment by Onyx2406** [comment]
> Code is good. However the commit message body contains some text that has little to no value. Once the commit message is fixed this PR can be merged.
> 
> Note that [std::string::length()](https://en.cppreference.com/w/cpp/string/basic_string/length) returns the size of the string in bytes rather than in characters, i.e. the checks will not work correctly on non-ascii texts. But that can be fixed in s separate PR.

Hi, which commit message body? Can you maybe give reference.

**Review by veloman-yunkan — COMMENTED** [review]
The title of the commit message is too long - it exceeds the recommended limit, hence part of the title is moved into the body. If you want to mention the new checks in the title, you better split your changes into two commits:

1. Introduce the checks on `--description` option
2. Add `--longDescription` option (along with its own checks, but the latter part doesn't need to be in the commit title)

**Comment by Onyx2406** [comment]
> The title of the commit message is too long - it exceeds the recommended limit, hence part of the title is moved into the body. If you want to mention the new checks in the title, you better split your changes into two commits:
> 
> 1. Introduce the checks on `--description` option
> 2. Add `--longDescription` option (along with its own checks, but the latter part doesn't need to be in the commit title)

Looks good now..

**Review by veloman-yunkan — APPROVED** [review]

File: src/zimwriterfs/zimwriterfs.cpp [hunk 5]
```diff
@@ -80,6 +81,28 @@ bool thereAreMissingArguments()
...
+    return false;
+  }
+
+  if (!longDescription.empty() && longDescription.length() > 4000) {
+    std::cerr << "Long description length exceeds the 4000 character limit." << std::endl;
+    return false;
+  }
+
+  return true;
 }
```
Comment (veloman-yunkan) [inline src/zimwriterfs/zimwriterfs.cpp:106]: Missing closing brace

**Comment by veloman-yunkan** [comment]
> @veloman-yunkan @Onyx2406 Part of the CI does not pass, it seems to me we have a regression here.

Indeed. I missed a missing closing brace during review since I trust the compiler. Looks like @Onyx2406 didn't build the latest code before requesting review.

**Comment by mgautierfr** [comment]
> @kelson42 BTW, I wonder why the list of CI jobs in this PR is limited to Packages workflows.

This is because the PR is coming from a branch in another repository. And the CI workflow is not run for security reasons.
However, we may have to check the exact status about that and do something to homogenize all CIs.

**Pull request merged**